src/thunderbots/software/ai/hl/stp/tactic/tactic.cpp

```diff
@@ -11,18 +13,46 @@ bool Tactic::done() const
     // If the coroutine "iterator" is done (ie. evaluates to false, has no more values
     // to iterate), the calculateNextIntent function has completed and therefore
     // the Tactic is done
-    return !static_cast<bool>(intent_sequence);
+    //
+    // We also check out internal done_ variable. This is set automatically by the
+    // getNextIntent function if we ever return a nullptr. We do this extra check because
+    // Tactics can return a nullptr one "tick" (aka function call) before the coroutine
+    // actually evaluates as done. This way we ensure that the Tactic::done() function
+    // will always evaluate to true on the exact same call a nullptr is returned.
+    //
+    // This is important since higher-level functions rely on the done() function, and
+    // returning nullptr values "out of sync" with this could cause problems
+    return done_ || !static_cast<bool>(intent_sequence);
```

Comment:
Why do we check `!static_cast<bool>(intent_sequence)` here if `done_` is already set?

Reply:
The coroutine could complete without a nullptr being returned first, in which case the `done_` variable wouldn't be set.

Alternatively we can also set the `done_` variable after the `if(couroutine)` check anyway (where we default to returning a nullptr at the end of the function.

```
getNextIntent() {
    if(coroutine) {
        // Set done_ to true if a nullptr was returned
        // return result
    }
    done_ = true;
    return nullptr;
}
```

Reply:
Interesting, what's the difference between returning the nullptr and just completing? What would be the case for both? 

Either way, I think what you just suggested (setting `done_` in all cases) is a good option!

Reply:
There's not much of a difference between them (both indicate we are "done"), but implementation details can cause a nullptr to be returned one iteration/tick before the `done()` function evaluates to `true`. We want these to always be perfectly in sync.

```
MoveAction move_action = MoveAction(*robot);
do
{
    yield(move_action.updateStateAndGetNextIntent(*robot, destination, final_orientation, final_speed));
} while (!move_action.done());
```

In this example, a nullptr will be returned one iteration before the coroutine is done. This is because once the action is done, one more `Action::getNextIntent` will be yielded before we see the action is done and finish. This is the case with `while` loops as well, because `Action::getNextIntent` has to be called a second time for the Action to realize it's done, so we will always return a nullptr one iteration before the coroutine finishes.

On the other hand, if the Tactic was directly constructing and yielding Intents (which is valid), then the coroutine would finish normally without prematurely returning any nullptrs.

Hope that makes things clear. Regardless, I've made the change I suggested above (setting the `done_` variable again and just using that)